dimagi/commcare-hq
# Make links in Web Apps more accessible

## Summary

https://dimagi-dev.atlassian.net/browse/USH-394

Make links in Web Apps more accessible.

This is the first change to improve Web Apps accessibility based on feedback from screen reader users. The overall plan for this work is described in [this document](https://docs.google.com/document/d/1fHNaVlaFl1xrbtPbMqj0YpByss1e0_hzCRlAUbVLboM/edit?usp=sharing), which includes links to the feedback from a vendor and links to accessibility standards and resources.

This first change makes navigation links in Web Apps implement the [WAI-ARIA](https://www.w3.org/TR/wai-aria/) role of "link", which typically involves 4 parts:

1. find the appropriate HTML element and add the role="link" attribute which informs screen readers to treat it as a link.
2. put the element into the tab order with the tabindex="0" attribute, so it can receive keyboard focus.
3. add a keyboard event handler to the Marionette view for the element which treats Enter key presses like mouse clicks.
4. add a label with either aria-labelledby or aria-label to tell a screen reader what text to read for the element.

Using aria-labelledby, which refers to another element by id, is preferable to aria-label, so I tried to use aria-labelledby in most cases. Using aria-labelledby with translated terms should read them after translation, which definitely seems better.

## Product Description

For users without a screen reader, this change enables keyboard navigation of Web Apps, using the Tab and Enter keys. Navigating by keyboard is one way to test that it works. No other changes should be evident without a screen reader.

### QA Plan

The QA plan will be check navigation of these pages by mouse to ensure there are no regressions and to test navigation by keyboard to see if the changes actually work.

[QA ticket](https://dimagi-dev.atlassian.net/browse/QA-2188)

### Safety story

WAI-ARIA attributes are designed to have no effect outside of accessibility tools, so they shouldn't cause any regressions.

The other changes are putting elements into the focus ring and handling key clicks on them, which should not affect existing users who navigate by mouse clicks.

### Rollback instructions
